src/server/identity/server/dex_token_handler.go

```diff
@@ -0,0 +1,172 @@
+			return
+		}
+		if len(ps) == 0 {
+			return
+		}
+		token, err := idTokenFromBytes(ctx, srw.b.Bytes())
+		if err != nil {
+			// TODO: this doesn't work
+			rw.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		if token.Email == "" {
+			log.Error(ctx, "failed to find email claim in ID token", zap.Error(err))
+			rw.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		for _, p := range ps {
+			defer p.Close()
+			u := &User{name: token.Email}
+			if _, err := p.FindUser(ctx, token.Email); err != nil {
+				if !errors.As(err, &errNotFound{}) {
+					log.Error(ctx, "failed to find user", zap.Error(err),
+						zap.String("user", token.Email))
+					rw.WriteHeader(http.StatusInternalServerError)
+					return
+				}
+				if _, err := p.CreateUser(ctx, &User{name: token.Email}); err != nil {
+					log.Error(ctx, "failed to provision user", zap.Error(err),
+						zap.String("user", token.Email))
+					rw.WriteHeader(http.StatusInternalServerError)
+					return
+				}
+			}
+			var gs []*Group
+			for _, grp := range token.Groups {
+				g, err := p.FindGroup(ctx, grp)
+				if err != nil {
+					if !errors.As(err, &errNotFound{}) {
+						log.Error(ctx, "failed to find group", zap.Error(err),
+							zap.String("group", grp))
+						rw.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+					if _, err := p.CreateGroup(ctx, &Group{name: grp}); err != nil {
+						log.Error(ctx, "failed to provision group", zap.Error(err),
+							zap.String("group", grp))
+						rw.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+				}
+				gs = append(gs, g)
+			}
+			if err := p.SetUserGroups(ctx, u, gs); err != nil {
+				log.Error(ctx, "failed to set user groups", zap.Error(err),
+					zap.String("user", token.Email),
+					zap.Any("groups", token.Groups))
+				rw.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+		}
+	}
+}
+
+func idTokenFromBytes(ctx context.Context, b []byte) (*idToken, error) {
+	resp := &tokenResp{}
+	if err := json.Unmarshal(b, resp); err != nil {
+		msg := "failed to unmarshal /token response"
+		log.Error(ctx, msg, zap.Error(err))
+		return nil, errors.Wrap(err, msg)
+	}
+	tok, err := b64.RawStdEncoding.DecodeString(jwtPayload(resp.IdToken))
+	if err != nil {
+		msg := "failed to base64 decode ID token"
+		log.Error(ctx, msg, zap.Error(err),
+			zap.String("token", resp.IdToken))
+		return nil, errors.Wrap(err, msg)
+	}
+	token := &idToken{}
+	if err := json.Unmarshal(tok, token); err != nil {
+		msg := "failed to unmarshal /token response"
+		log.Error(ctx, msg, zap.Error(err))
+		return nil, errors.Wrap(err, msg)
+	}
+	return token, nil
+}
+
+type saveResponseWriter struct {
+	b  *bytes.Buffer
+	rw *http.ResponseWriter
+}
+
+func (sr *saveResponseWriter) Write(b []byte) (int, error) {
+	if _, err := sr.b.Write(b); err != nil {
+		return 0, errors.Wrap(err, "write bytes to save response writer buffer")
+	}
+	w := *sr.rw
+	i, err := w.Write(b)
```

Comment:
I think "writing now" makes sense. The main thing this struct is supposed to achieve is to behave like the ResponseWriter it's wrapping, while recording the bytes written

Reply:
You can’t write the HTTP response body without writing the HTTP header, and you can’t write the HTTP header without emitting a status. So if you write the HTTP response at this point, you won’t be able to change the status later — it will already have been emitted as a 200 (or whatever).

Reply:
I think that your updates resolve this.